### PR TITLE
feat: Add `pulumi docs read` command for terminal documentation browsing

### PR DESCRIPTION
**Pull request by CamSoper**
## Summary (PR 1 of 4)

> **Review/merge order:** This is the first PR. Review and merge this before PRs 2–4.

This is the first in a series of 4 PRs splitting the `pulumi docs` command (originally #22381) into reviewable chunks.

Adds the core read pipeline for browsing Pulumi documentation in the terminal:

- `pulumi docs <path>` / `pulumi docs read <path>` — fetch and render a docs page
- Glamour-based terminal markdown rendering with code block borders and note boxes
- Language/OS chooser system for conditional content (`--language`, `--os`)
- Section extraction with `--toc` for table of contents navigation
- User preferences persistence (language, OS, last viewed page)
- `--raw` and `--json` output modes

### PR series

1. **`pulumi docs read`** (this PR) — core read pipeline
2. `pulumi docs search/sitemap/registry` — additional subcommands
3. Bundle system — registry API docs with caching
4. `pulumi docs browse` — interactive browse mode

## Test plan

- [x] `go build ./cmd/pulumi/docs/...` compiles
- [x] `go build ./cmd/pulumi/` compiles (full CLI)
- [x] `go test -count=1 ./cmd/pulumi/docs/...` passes
- [x] `golangci-lint run ./cmd/pulumi/docs/...` clean

### CODE REVIEW

File: pkg/cmd/pulumi/docs/chooser.go [hunk 1]
```diff
@@ -0,0 +1,448 @@
+// Copyright 2024, Pulumi Corporation.
```
Comment (tgummerer) [inline pkg/cmd/pulumi/docs/chooser.go]: It's 2026.  Please look at the other files for this too.

File: pkg/cmd/pulumi/docs/chooser.go [hunk 2]
```diff
@@ -0,0 +1,448 @@
...
+		prefs.Set(c.Type, selection)
+		// Best-effort save
+		_ = prefs.Save()
+	}
+
+	// Build output
+	isInline := c.Prefix != "" || c.Suffix != ""
+
+	if selection != "" {
+		// Find the matching option
```
Comment (tgummerer) [inline pkg/cmd/pulumi/docs/chooser.go]: Please remove all the unnecessary comments like this one.  They just make the code harder to read without adding anything of value.  A  good comment should explain *why* we do something, not just a descripiton of what happens in the next lines, especially if that's obvious from the code.

File: pkg/cmd/pulumi/docs/chooser.go [hunk 3]
```diff
@@ -0,0 +1,448 @@
...
+			cmdutil.GetGlobalColorization(),
+		)
+	}
+
+	// Save selection
+	if selection != "" {
+		session[c.Type] = selection
+		prefs.Set(c.Type, selection)
+		// Best-effort save
+		_ = prefs.Save()
```
Comment (tgummerer) [inline pkg/cmd/pulumi/docs/chooser.go]: What happens if we fail here?

File: pkg/cmd/pulumi/docs/docs.go [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+	toc             bool
+	jsonOutput      bool
+}
+
+// NewDocsCmd creates the `pulumi docs` command.
+func NewDocsCmd() *cobra.Command {
+	dc := &docsCmd{}
+
+	cmd := &cobra.Command{
+		Use:   "docs [path[#section]]",
```
Comment (tgummerer) [inline pkg/cmd/pulumi/docs/docs.go]: This doesn't seem right, especially as we have subcommands here.  Also the constrictor library overwrites this anyway iirc, so just keep it as


```suggestion
		Use:   "docs",
```

File: pkg/cmd/pulumi/docs/docs.go [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+		Short: "View Pulumi documentation in the terminal",
+		Long: "Read and browse Pulumi documentation in the terminal.\n\n" +
+			"  pulumi docs                    Show help\n" +
+			"  pulumi docs <path>             Read a specific page\n" +
+			"  pulumi docs read <path>        Read a specific page",
```
Comment (tgummerer) [inline pkg/cmd/pulumi/docs/docs.go]: Why do both `docs <path>` and `docs read <path>` work here?  Imo, we should only allow `pulumi docs read <path>` for a start, otherwise we just have to deal with more ambiguity.

File: pkg/cmd/pulumi/docs/docs.go [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+			if len(args) > 0 {
+				// Bare path shortcut → read
+				return dc.fetchAndRender(args[0])
+			}
+			return cmd.Help()
+		},
+	}
+
+	dc.baseURL = "https://www.pulumi.com"
+	if envURL := os.Getenv("PULUMI_DOCS_BASE_URL"); envURL != "" {
```
Comment (tgummerer) [inline pkg/cmd/pulumi/docs/docs.go]: Instead of env vars, how about making these command line flags? Then you also get env vars for free, and there's no particular advantage for these being only env vars I think.

File: pkg/cmd/pulumi/docs/docs.go [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+				return cmd.Help()
+			}
+			return dc.fetchAndRender(args[0])
+		},
+	}
+	cmd.Flags().BoolVar(&dc.raw, "raw", false,
+		"Output raw markdown without formatting or chooser resolution")
+	cmd.Flags().BoolVar(&dc.toc, "toc", false,
+		"Show table of contents (list of sections)")
+	cmd.Flags().BoolVar(&dc.jsonOutput, "json", false,
```
Comment (tgummerer) [inline pkg/cmd/pulumi/docs/docs.go]: This doesn't work:

```
$ pulumi docs read iac/concepts/stacks --json | head
# Stacks

Every Pulumi program is deployed to a _stack_. A stack is an isolated, independently [configurable](/docs/concepts/config/) instance of a Pulumi program. Stacks are commonly used to denote different phases of development (such as `development`, `staging`, and `production`) or feature branches (such as `feature-x-dev`).

A project can have as many stacks as you need. By default, Pulumi creates a stack for you when you start a new project using the `pulumi new` command.

## Create a stack {#create-stack}

To create a new stack, use `pulumi stack init stackName`. This creates an empty stack `stackName` and sets it as the *active* stack. The project that the stack is associated with is determined by finding the nearest `Pulumi.yaml` file.
```

that's not json.

File: pkg/cmd/pulumi/pulumi.go [hunk 8]
```diff
@@ -56,6 +56,7 @@ import (
 	"github.com/pulumi/pulumi/pkg/v3/cmd/pulumi/console"
 	"github.com/pulumi/pulumi/pkg/v3/cmd/pulumi/convert"
 	"github.com/pulumi/pulumi/pkg/v3/cmd/pulumi/deployment"
+	docsCmd "github.com/pulumi/pulumi/pkg/v3/cmd/pulumi/docs"
```
Comment (tgummerer) [inline pkg/cmd/pulumi/pulumi.go]: Why the rename?

File: pkg/cmd/pulumi/docs/sections_test.go [hunk 9]
```diff
@@ -0,0 +1,288 @@
...
+	})
+
+	t.Run("mixed content", func(t *testing.T) {
+		t.Parallel()
+		md := "See [Stacks](/docs/stacks) and [Google](https://google.com)."
+		result := stripExternalLinks(md)
+		assert.Contains(t, result, "[Stacks](/docs/stacks)")
+		assert.NotContains(t, result, "https://google.com")
+		assert.Contains(t, result, "Google")
+	})
```
Comment (tgummerer) [inline pkg/cmd/pulumi/docs/sections_test.go]: No need for 3 tests here, the last one shows that it all works correctly.  Also why not just use `assert.Equal` on the whole text in the last test?

File: pkg/cmd/pulumi/docs/chooser.go (hunk 1 above)
Comment (CamSoper) [reply pkg/cmd/pulumi/docs/chooser.go]: I mean, sure, but to be clear, this was not arbitrary. I looked at other recently changed files to see what the boilerplate was. I saw 2024 so I assumed there must be a reason for it. I'll update it.

File: pkg/cmd/pulumi/docs/docs.go (hunk 6 above)
Comment (CamSoper) [reply pkg/cmd/pulumi/docs/docs.go]: I originally had them as command line flags, but I couldn't think of any reason in the world why anybody but me (or some future maintainer) would ever have any reason to use these. They're exclusively used for testing/debugging purposes, so I didn't want to clutter the help text with information that's not helpful for the user.

File: pkg/cmd/pulumi/docs/docs.go (hunk 7 above)
Comment (CamSoper) [reply pkg/cmd/pulumi/docs/docs.go]: Nor is it meant to be.

```output
Flags:
  -h, --help   help for read
      --json   Output as JSON (use with --toc for structured section list)
      --raw    Output raw markdown without formatting or chooser resolution
      --toc    Show table of contents (list of sections)
```

You didn't use it with `--toc`. But I see the usability concern, I'll address it by renaming this one to `--toc-json`

File: pkg/cmd/pulumi/docs/chooser.go (hunk 1 above)
Comment (CamSoper) [reply pkg/cmd/pulumi/docs/chooser.go]: Fixed, updated all files to 2026.

File: pkg/cmd/pulumi/docs/chooser.go (hunk 2 above)
Comment (CamSoper) [reply pkg/cmd/pulumi/docs/chooser.go]: Agreed, cleaned them up. Kept only the one explaining why session selections are reused across choosers.

File: pkg/cmd/pulumi/docs/chooser.go (hunk 3 above)
Comment (CamSoper) [reply pkg/cmd/pulumi/docs/chooser.go]: This saves UI preferences (last language/OS choice) to `~/.pulumi/docs-preferences.json`. If the write fails, the user just gets re-prompted next time — propagating the error here would abort the entire page render over a non-critical preference save. Added a `logging.V(7)` debug log so failures are visible when troubleshooting.

File: pkg/cmd/pulumi/docs/docs.go (hunk 4 above)
Comment (CamSoper) [reply pkg/cmd/pulumi/docs/docs.go]: Fixed.

File: pkg/cmd/pulumi/docs/docs.go (hunk 5 above)
Comment (CamSoper) [reply pkg/cmd/pulumi/docs/docs.go]: Good call, removed the bare path shortcut. `pulumi docs read <path>` only now.

File: pkg/cmd/pulumi/docs/docs.go (hunk 6 above)
Comment (CamSoper) [reply pkg/cmd/pulumi/docs/docs.go]: Done — `--base-url` and `--registry-base-url` persistent flags, with the env vars as defaults.

File: pkg/cmd/pulumi/docs/docs.go (hunk 7 above)
Comment (CamSoper) [reply pkg/cmd/pulumi/docs/docs.go]: Renamed to `--toc-json` so the scope is self-documenting. It also implies `--toc` now so you don't need both flags.

File: pkg/cmd/pulumi/pulumi.go (hunk 8 above)
Comment (CamSoper) [reply pkg/cmd/pulumi/pulumi.go]: Dropped the alias, no conflict with a bare `docs` import.

File: pkg/cmd/pulumi/docs/sections_test.go (hunk 9 above)
Comment (CamSoper) [reply pkg/cmd/pulumi/docs/sections_test.go]: Collapsed to one test with `assert.Equal` on the full output.

File: pkg/cmd/pulumi/docs/chooser.go [hunk 10]
```diff
@@ -0,0 +1,414 @@
...
+		content := body[contentStart : contentStart+closeLoc[0]]
+		content = strings.TrimLeft(content, "\n")
+		content = strings.TrimRight(content, "\n")
+
+		chooser.Options = append(chooser.Options, Option{
+			Value:   optValue,
+			Content: content,
+		})
+
+		pos = contentStart + closeLoc[1]
```
Comment (tgummerer) [inline pkg/cmd/pulumi/docs/chooser.go:163]: Can't we use regexes to extract this instead of all this open coded stuff, using capture groups?

If the comment is right, something like

```
re := regexp.MustParse(`<!-- option: (\w+) -->(.*?)<!-- /option -->`)
submatches := re.FindAllSubmatch(body)
for submatch := range submatches {
  chooser.Options = append(chooser.Options, Option{ Value: submatch[1], Content: submatch[2])
```

Same question for the choosers as well.

File: pkg/cmd/pulumi/docs/chooser.go [hunk 11]
```diff
@@ -0,0 +1,414 @@
...
+
+// stripLeftoverTags removes lines that consist entirely of a chooser/option HTML comment tag.
+// It does NOT modify lines that mix tag and non-tag content to avoid mangling the output.
+func stripLeftoverTags(s string) string {
+	lines := strings.Split(s, "\n")
+	var cleaned []string
+	for _, line := range lines {
+		if chooserTagRe.MatchString(line) {
+			continue
+		}
```
Comment (tgummerer) [inline pkg/cmd/pulumi/docs/chooser.go:197]: So this is for just cleaning up leftover tags that are *wrong* in the documentation I assume?  So two questions:
1) can't we just clean them up in the documentation?
2) if we can't, should we really drop the whole line instead of just the chooser?

File: pkg/cmd/pulumi/docs/chooser_test.go [hunk 12]
```diff
@@ -0,0 +1,281 @@
...
+}
+
+func TestFilterCodeBlocksByLanguage(t *testing.T) {
+	t.Parallel()
+
+	multiLangExample := "## Example Usage\n\n" +
+		"```typescript\nconsole.log('hello');\n```\n\n\n\n" +
+		"```python\nprint('hello')\n```\n\n\n\n" +
+		"```go\nfmt.Println(\"hello\")\n```\n\n" +
+		"## Next Section\n"
```
Comment (tgummerer) [inline pkg/cmd/pulumi/docs/chooser_test.go]: Please format this with the newlines being actual newlines, this is very hard to read.

File: pkg/cmd/pulumi/docs/chooser_test.go [hunk 13]
```diff
@@ -0,0 +1,281 @@
...
+	})
+
+	t.Run("unknown type returns empty", func(t *testing.T) {
+		t.Parallel()
+		p := &Preferences{Language: "go"}
+		assert.Equal(t, "", p.Get("unknown"))
+	})
+}
+
+func TestFilterCodeBlocksByLanguage(t *testing.T) {
```
Comment (tgummerer) [inline pkg/cmd/pulumi/docs/chooser_test.go:231]: This needs a bunch more tests. Just tests for different languages are not all that helpful, but we probably want to test misformed code blocks, code blocks that have no language, etc. This `TestFilterCodeBlocks` function looks very complex, so we should also be testing all that complexity (or reduce the complexity of the function if we can ideally).

File: pkg/cmd/pulumi/docs/docs.go [hunk 14]
```diff
@@ -0,0 +1,277 @@
...
+	dc := &docsCmd{}
+
+	defaultBaseURL := "https://www.pulumi.com"
+	if envURL := os.Getenv("PULUMI_DOCS_BASE_URL"); envURL != "" {
+		defaultBaseURL = envURL
+	}
+	defaultRegistryURL := defaultBaseURL
+	if envURL := os.Getenv("PULUMI_REGISTRY_BASE_URL"); envURL != "" {
+		defaultRegistryURL = envURL
+	}
```
Comment (tgummerer) [inline pkg/cmd/pulumi/docs/docs.go]: We don't need these. Just the the default, and every flag automatically gets a command line flag `PULUMI_OPTION_<flagname>` by default.

File: pkg/cmd/pulumi/docs/docs.go [hunk 15]
```diff
@@ -0,0 +1,277 @@
...
+		"Filter OS-specific content in docs (e.g., macos, linux, windows); choice is remembered")
+
+	cmd.AddCommand(dc.newReadCmd())
+
+	return cmd
+}
+
+func (dc *docsCmd) newReadCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "read <path[#section]>",
```
Comment (tgummerer) [inline pkg/cmd/pulumi/docs/docs.go:68]: Again, the usage string is taken from the constrictor, so this won't actually show up in any output for the user. It currently shows up as:

``1
Usage:
  pulumi docs read [path] [flags]
```

I think we probably want the `#section` in there, which you can do using the `Usage` field in `constrictor.Argument`.

File: pkg/cmd/pulumi/docs/docs.go [hunk 16]
```diff
@@ -0,0 +1,277 @@
...
+			"  pulumi docs read iac/concepts/stacks#stack-tags    Jump to a section\n" +
+			"  pulumi docs read registry/packages/aws             Read a registry page\n" +
+			"  pulumi docs read --toc                             Show sections on last viewed page",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if dc.tocJSON {
+				dc.toc = true
+			}
+			if dc.toc && len(args) == 0 {
+				prefs, _ := LoadPreferences()
+				if prefs.LastPage == "" {
```
Comment (tgummerer) [inline pkg/cmd/pulumi/docs/docs.go:81]: Imo this last page preference is very surprising.  Shouldn't `pulumi docs read` without arguments show you the "docs home" from pulumi.com/docs instead?  Or a table of contents?  It's not likely someone using this command the first time knows which pages to visit, so some sort of index would be nice.

File: pkg/cmd/pulumi/docs/chooser.go (hunk 10 above)
Comment (CamSoper) [reply pkg/cmd/pulumi/docs/chooser.go:163]: The position-based approach in `ParseChoosers` is needed because it detects inline vs block choosers (single-line choosers that have prefix/suffix text on the same line, like `Install <!-- chooser: os --><!-- option: macos -->Homebrew<!-- /option --><!-- option: linux -->apt<!-- /option --><!-- /chooser --> and run it`). A regex can't extract the surrounding prefix/suffix context from the same line.

`parseOptions` is already 25 lines — it just walks open/close tag pairs and extracts the multi-line content between them. A `(?s)` regex with `(.*?)` for multi-line capture gets fragile with nested content and is harder to debug when it breaks. Happy to add a comment explaining why position-based parsing is used here.

File: pkg/cmd/pulumi/docs/chooser.go (hunk 11 above)
Comment (CamSoper) [reply pkg/cmd/pulumi/docs/chooser.go:197]: These tags aren't bugs in the docs — they're the mechanism by which pulumi.com renders language/OS choosers in the browser. The web JS processes `<!-- chooser: language -->` / `<!-- /chooser -->` to build interactive selectors. We resolve the chooser to the user's selection and then strip the now-redundant comment tags so they don't appear as raw text in the terminal.

On your second question: the function already only drops lines that consist *entirely* of a tag. Lines with mixed content (tag + real text on the same line) are preserved — that's what the regex anchoring with `^\s*` and `\s*$` enforces.

File: pkg/cmd/pulumi/docs/chooser_test.go (hunk 12 above)
Comment (CamSoper) [reply pkg/cmd/pulumi/docs/chooser_test.go]: Fixed — using raw string literals now.

File: pkg/cmd/pulumi/docs/chooser_test.go (hunk 13 above)
Comment (CamSoper) [reply pkg/cmd/pulumi/docs/chooser_test.go:231]: Added edge case tests: empty input, code block with no language tag, single code block (no group to filter), and unclosed fence.

File: pkg/cmd/pulumi/docs/docs.go (hunk 14 above)
Comment (CamSoper) [reply pkg/cmd/pulumi/docs/docs.go]: TIL about `declareFlagsAsEnvironmentVariables` — removed the manual `os.Getenv` reads. The auto-generated `PULUMI_OPTION_BASE_URL` / `PULUMI_OPTION_REGISTRY_BASE_URL` handle it.

File: pkg/cmd/pulumi/docs/docs.go (hunk 15 above)
Comment (CamSoper) [reply pkg/cmd/pulumi/docs/docs.go:68]: Good call — updated to `constrictor.Argument{Name: "path", Usage: "path[#section]"}`. Help now shows `pulumi docs read path[#section] [flags]`.

File: pkg/cmd/pulumi/docs/docs.go (hunk 16 above)
Comment (CamSoper) [reply pkg/cmd/pulumi/docs/docs.go:81]: Agreed the first-time UX should be better. For this PR, `pulumi docs read` with no args shows help — the last-page shortcut only kicks in with `--toc` (so you can quickly check sections of what you just read). The discovery/index experience you're describing is exactly what `pulumi docs browse` provides in PR #22401.

File: pkg/cmd/pulumi/docs/docs.go [hunk 17]
```diff
@@ -0,0 +1,267 @@
...
+	language        string
+	osFlag          string
+	raw             bool
+	toc             bool
+	tocJSON         bool
+}
+
+// NewDocsCmd creates the `pulumi docs` command.
+func NewDocsCmd() *cobra.Command {
+	dc := &docsCmd{}
```
Comment (iwahbe) [inline pkg/cmd/pulumi/docs/docs.go:40]: I'd call this docs args

File: pkg/cmd/pulumi/docs/docs.go [hunk 18]
```diff
@@ -0,0 +1,267 @@
...
+		Long: "Read and browse Pulumi documentation in the terminal.\n\n" +
+			"  pulumi docs                    Show help\n" +
+			"  pulumi docs read <path>        Read a specific page",
+	}
+
+	cmd.PersistentFlags().StringVar(&dc.baseURL, "base-url", "https://www.pulumi.com",
+		"Base URL for Pulumi documentation")
+	cmd.PersistentFlags().StringVar(&dc.registryBaseURL, "registry-base-url", "https://www.pulumi.com",
+		"Base URL for Pulumi registry")
+	cmd.PersistentFlags().MarkHidden("base-url")          //nolint:errcheck
```
Comment (iwahbe) [inline pkg/cmd/pulumi/docs/docs.go:54]: Let's use `contract.AssertNoErrorf` here instead.

File: pkg/cmd/pulumi/docs/docs.go [hunk 19]
```diff
@@ -0,0 +1,267 @@
...
+	}
+	cmd.Flags().BoolVar(&dc.raw, "raw", false,
+		"Output raw markdown without formatting or chooser resolution")
+	cmd.Flags().BoolVar(&dc.toc, "toc", false,
+		"Show table of contents (list of sections)")
+	cmd.Flags().BoolVar(&dc.tocJSON, "toc-json", false,
+		"Output table of contents as JSON")
+	constrictor.AttachArguments(cmd, &constrictor.Arguments{
+		Arguments: []constrictor.Argument{{Name: "path", Usage: "path[#section]"}},
+	})
```
Comment (iwahbe) [inline pkg/cmd/pulumi/docs/docs.go:100]: The scoping of `dc` feels wrong here. Does `docsCmd` represent the args of a docs command, or the args of the `read` docs command?

File: pkg/cmd/pulumi/docs/docs.go [hunk 20]
```diff
@@ -0,0 +1,267 @@
...
+			}
+			return dc.fetchAndRender(args[0])
+		},
+	}
+	cmd.Flags().BoolVar(&dc.raw, "raw", false,
+		"Output raw markdown without formatting or chooser resolution")
+	cmd.Flags().BoolVar(&dc.toc, "toc", false,
+		"Show table of contents (list of sections)")
+	cmd.Flags().BoolVar(&dc.tocJSON, "toc-json", false,
+		"Output table of contents as JSON")
```
Comment (iwahbe) [inline pkg/cmd/pulumi/docs/docs.go:97]: Does this have an implied "and exit"? If so, let's make it explicit.

File: pkg/cmd/pulumi/docs/docs.go [hunk 21]
```diff
@@ -0,0 +1,267 @@
...
+			if len(args) == 0 {
+				return cmd.Help()
+			}
+			return dc.fetchAndRender(args[0])
+		},
+	}
+	cmd.Flags().BoolVar(&dc.raw, "raw", false,
+		"Output raw markdown without formatting or chooser resolution")
+	cmd.Flags().BoolVar(&dc.toc, "toc", false,
+		"Show table of contents (list of sections)")
```
Comment (iwahbe) [inline pkg/cmd/pulumi/docs/docs.go:95]: Asking because of `toc-json`, does this have an implied "and exit"?

File: pkg/cmd/pulumi/docs/sections.go [hunk 22]
```diff
@@ -0,0 +1,225 @@
...
+
+var linkRe = regexp.MustCompile(`\[([^\]\n]+)\]\((/(docs|registry)/[^\)\n]+)\)`)
+
+// allLinkRe matches any markdown link (single-line only).
+var allLinkRe = regexp.MustCompile(`\[([^\]\n]+)\]\([^)\n]+\)`)
+
+// stripExternalLinks replaces non-internal markdown links with just their
+// display text. Internal links (/docs/... and /registry/...) are preserved.
+// This prevents glamour from rendering unhelpful raw URLs inline.
+func stripExternalLinks(md string) string {
```
Comment (iwahbe) [inline pkg/cmd/pulumi/docs/sections.go:31]: I'm not sure we want this. These links seem helpful.

File: pkg/cmd/pulumi/docs/chooser.go
Comment (iwahbe) [inline pkg/cmd/pulumi/docs/chooser.go:1]: We already have `codegen.FilterExamples`, can we use that?

https://github.com/pulumi/pulumi/blob/655c2d9674f558682a84508785822967e6c6d382/pkg/codegen/docs.go#L138-L149

File: pkg/cmd/pulumi/docs/sections.go [hunk 23]
```diff
@@ -0,0 +1,225 @@
...
+
+// heading represents a markdown heading with its nesting level and URL slug.
+type heading struct {
+	level int
+	text  string
+	slug  string
+}
+
+// extractIntro returns the content before the first ## heading.
+func extractIntro(md string) string {
```
Comment (iwahbe) [inline pkg/cmd/pulumi/docs/sections.go:106]: This is never used!

File: pkg/cmd/pulumi/docs/sections.go [hunk 24]
```diff
@@ -0,0 +1,225 @@
...
+			// Only one section — return the whole thing.
+			return md
+		}
+	}
+	return md
+}
+
+// introContainsFirstHeading returns true if the intro (as returned by extractIntro)
+// includes the first ## heading because there was no text before it.
+func introContainsFirstHeading(md string) bool {
```
Comment (iwahbe) [inline pkg/cmd/pulumi/docs/sections.go:130]: This is also never used!

**Review by iwahbe — COMMENTED** [review]
I'd like to see more structure here. At minimum, we should have 1 package with a simple interface that takes some settings and a string, and renders it out.

The command should consume that package & do the thing.

**Comment by CamSoper** [comment]
Closing this in favor of a cleaner rewrite. The review feedback from @tgummerer and @iwahbe was helpful — the key takeaways:

- **Use goldmark for markdown parsing** instead of regex/string munging. The codebase already depends on `github.com/pgavlin/goldmark` and has patterns for custom AST nodes (see `pkg/codegen/schema/docs_parser.go`). Chooser resolution, heading extraction, section slicing, and code block filtering should all operate on the AST.
- **Separate library from CLI.** A `pkg/docsrender/` package with a clean interface (settings + content in, rendered output out), consumed by thin cobra commands in `pkg/cmd/pulumi/docs/`.
- **Reuse `codegen.FilterExamples`** (or its approach) for language filtering.
- **Scoped args structs** — don't share one mutable struct across subcommands with different flag sets.

The new implementation will target a feature branch with smaller, independently reviewable PRs. Same functionality, better bones.
